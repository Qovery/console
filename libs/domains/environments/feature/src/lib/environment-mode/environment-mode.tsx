import { type VariantProps, cva } from 'class-variance-authority'
import { type EnvironmentModeEnum } from 'qovery-typescript-axios'
import { type ElementRef, forwardRef } from 'react'
import { match } from 'ts-pattern'
import { Badge, type BadgeProps } from '@qovery/shared/ui'
import { twMerge } from '@qovery/shared/util-js'

const environmentModeVariants = cva('', {
  variants: {
    variant: {
      shrink: ['flex', 'h-4', 'w-4', 'justify-center', 'p-0', 'font-semibold'],
      full: [],
    },
  },
  defaultVariants: {
    variant: 'full',
  },
})

export interface EnvironmentModeProps
  extends Omit<BadgeProps, 'color' | 'variant'>,
    VariantProps<typeof environmentModeVariants> {
  mode: keyof typeof EnvironmentModeEnum
}

export const EnvironmentMode = forwardRef<ElementRef<typeof Badge>, EnvironmentModeProps>(
  ({ mode, className, variant, ...props }, forwardedRef) => {
    variant ??= 'full'
    return match(mode)
      .with('PRODUCTION', () => (
        <Badge
          ref={forwardedRef}
          variant="outline"
          color="red"
          className={twMerge(className, environmentModeVariants({ variant }))}
          {...props}
        >
          {variant === 'full' ? 'Production' : 'P'}
        </Badge>
      ))
      .with('DEVELOPMENT', () => (
        <Badge
          ref={forwardedRef}
          variant="outline"
          color="neutral"
          className={twMerge(className, environmentModeVariants({ variant }))}
          {...props}
        >
          {variant === 'full' ? 'Development' : 'D'}
        </Badge>
      ))
      .with('PREVIEW', () => (
        <Badge
          ref={forwardedRef}
          variant="outline"
          color="purple"
          className={twMerge(className, environmentModeVariants({ variant }))}
          {...props}
        >
          {variant === 'full' ? 'Preview' : 'V'}
        </Badge>
      ))
      .with('STAGING', () => (
        <Badge
          ref={forwardedRef}
          variant="outline"
          color="neutral"
          className={twMerge(className, environmentModeVariants({ variant }))}
          {...props}
        >
          {variant === 'full' ? 'Staging' : 'S'}
        </Badge>
      ))
      .exhaustive()
  }
)

export default EnvironmentMode
