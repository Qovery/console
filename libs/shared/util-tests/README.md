# shared-util-tests

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test shared-util-tests` to execute the unit tests via [Jest](https://jestjs.io).
