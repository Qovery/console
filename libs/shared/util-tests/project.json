{
  "name": "shared-util-tests",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util-tests/src",
  "projectType": "library",
  "tags": ["scope:shared", "type:util"],
  "implicitDependencies": ["!shared-ui"],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
