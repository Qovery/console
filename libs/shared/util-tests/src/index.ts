export * from './lib/shared-util-tests'
