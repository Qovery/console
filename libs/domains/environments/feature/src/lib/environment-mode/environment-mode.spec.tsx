import { renderWithProviders } from '@qovery/shared/util-tests'
import { EnvironmentMode } from './environment-mode'

describe('EnvironmentMode', () => {
  it('should render as production mode', () => {
    const { baseElement } = renderWithProviders(<EnvironmentMode mode="PRODUCTION" />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should render as staging mode', () => {
    const { baseElement } = renderWithProviders(<EnvironmentMode mode="STAGING" />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should render as preview mode', () => {
    const { baseElement } = renderWithProviders(<EnvironmentMode mode="PREVIEW" />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should render as development mode', () => {
    const { baseElement } = renderWithProviders(<EnvironmentMode mode="DEVELOPMENT" />)
    expect(baseElement).toMatchSnapshot()
  })
})
