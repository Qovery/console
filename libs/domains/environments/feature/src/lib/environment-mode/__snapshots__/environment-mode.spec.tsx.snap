// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvironmentMode should render as development mode 1`] = `
<body>
  <div>
    <span
      class="inline-flex items-center shrink-0 font-medium text-xs h-6 px-2 border rounded border-neutral-250 text-neutral-350 dark:border-neutral-350 dark:text-neutral-300"
      data-accent-color="neutral"
    >
      Development
    </span>
  </div>
</body>
`;

exports[`EnvironmentMode should render as preview mode 1`] = `
<body>
  <div>
    <span
      class="inline-flex items-center shrink-0 font-medium text-xs h-6 px-2 text-purple-500 border rounded border-purple-500"
      data-accent-color="purple"
    >
      Preview
    </span>
  </div>
</body>
`;

exports[`EnvironmentMode should render as production mode 1`] = `
<body>
  <div>
    <span
      class="inline-flex items-center shrink-0 font-medium text-xs h-6 px-2 text-red-500 border rounded border-red-500"
      data-accent-color="red"
    >
      Production
    </span>
  </div>
</body>
`;

exports[`EnvironmentMode should render as staging mode 1`] = `
<body>
  <div>
    <span
      class="inline-flex items-center shrink-0 font-medium text-xs h-6 px-2 border rounded border-neutral-250 text-neutral-350 dark:border-neutral-350 dark:text-neutral-300"
      data-accent-color="neutral"
    >
      Staging
    </span>
  </div>
</body>
`;
